core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -76,13 +82,49 @@ class ReplicaAlterLogDirsThread(name: String,
     futureLog.updateHighWatermark(partitionData.highWatermark)
     futureLog.maybeIncrementLogStartOffset(partitionData.logStartOffset, LogStartOffsetIncrementReason.LeaderOffsetIncremented)
 
-    if (partition.maybeReplaceCurrentWithFutureReplica())
-      removePartitions(Set(topicPartition))
+    directoryEventHandler match {
+      case DirectoryEventHandler.NOOP =>
+        if (partition.maybeReplaceCurrentWithFutureReplica())
+          removePartitions(Set(topicPartition))
+      case _ =>
+        maybePromoteFutureReplica(topicPartition, partition)
+    }
 
     quota.record(records.sizeInBytes)
     logAppendInfo
   }
 
+  // Visible for testing
+  def updatedAssignmentRequestStat(topicPartition: TopicPartition)(state: DirectoryEventRequestState): Unit = {
+    assignmentRequestStates.put(topicPartition, state)
+  }
+  private def maybePromoteFutureReplica(topicPartition: TopicPartition, partition: Partition) = {
+    val partitionRequestState = Option(assignmentRequestStates.get(topicPartition))
+    val topicId = partition.topicId
+    if (topicId.isEmpty)
+      throw new IllegalStateException(s"Topic ${topicPartition.topic()} exists but its ID doesn't exist.")
+
+    partitionRequestState match {
+      case None =>
+        // Schedule assignment request and don't promote the future replica yet until the controller accepted the request.
+        partition.maybeFutureReplicaCaughtUp(_ => {
+          partition.futureReplicaDirectoryId()
+            .map {
+              directoryEventHandler.handleAssignment(new TopicIdPartition(topicId.get, topicPartition.partition()), _,
+                updatedAssignmentRequestStat(topicPartition)(_))
+            }
+        })
+      case Some(DirectoryEventRequestState.COMPLETED) =>
+        // Promote future replica if controller accepted the request and the replica caught-up with the original log.
+        if (partition.maybeReplaceCurrentWithFutureReplica()) {
+          removePartitions(Set(topicPartition))
+          assignmentRequestStates.remove(topicPartition)
+        }
+      case _ =>
```

Comment:
It seems we don't actually care about `DISPATCHED`, maybe we don't need that state.

Reply:
I updated the pr to only deal with `QUEUED` and `COMPLETED`. And dropped `DISPATCHED`

Reply:
Looks better. Thanks